### ISSUE TITLE
Range operations with undefined behavior

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#17123]
Currently several methods on `range` have "unspecified" behavior when they cannot provide a definitive answer, for example:

> proc range.high: Return the range’s high bound. If the range does not have a high bound the behavior is undefined.

At the same time `range.isEmpty()` "reports an error" (read: it halts) when there is no good answer.

This issue asks whether this is a good design or whether we want to change the specification (and implementation).

**Comment by lydia-duncan** [comment]
It seems to me like the determination of whether this is appropriate varies on a case by case basis.  I wouldn't lose sleep if the range type has some methods with undefined behavior and some where the bad case results in an error being thrown or a halt or whatever, but specific instances or comparisons may give me more pause (for instance, if `high` has undefined behavior but `low` does not, that seems bad)

**Comment by bradcray** [comment]
In reviewing this, I'm having trouble finding a case that gives undefined behavior.  For example, `.high` seems to give a compiler error if the high bound is not well-defined.  Where did the quote above come from, Vass?

**Comment by bradcray** [comment]
> when we learn that only at run time -- assuming it can happen, for example, when the alignment is undefined

I don't think this can ever happen, currently.  Even when alignment is undefined, say for `..n by 2`, we still know whether the high bound is known or not at compile-time (it's `n` in this example, despite the unknown alignment).

Note that `last` or `alignedHigh` are potentially not knowable at compile-time (so if/when [`high` becomes `alignedHigh`](https://github.com/chapel-lang/chapel/issues/17130), Vass's concern may become a reality.  But maybe we can close this issue in the meantime and deal with that case as part of that effort?

**Comment by bradcray** [comment]
> But maybe we can close this issue in the meantime and deal with that case as part of that effort?

I ended up addressing this in #18315 and am not aware of other cases where range operations have undefined behavior (and it certainly seems like the case in the OP doesn't), so am going to close this issue.  If/when we find cases that do have this issue, we can re-purpose this issue or open a new one.
